datafusion-physical-expr/src/expressions/case.rs

```diff
@@ -523,6 +528,39 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn case_with_expr_divide_by_zero() -> Result<()> {
+        let batch = case_test_batch1()?;
+        let schema = batch.schema();
+
+        // CASE a when 0 THEN float64(null) ELSE 25.0 / cast(a, float64)  END
+        let when1 = lit(ScalarValue::Int32(Some(0)));
```

Comment:
This test would fail since we are currently evaluating `else` first for `CASE expr WHEN value THEN result ELSE result END`, regardless of whether the when branches are entered.

Reply:
Currently, we are evaluating `case_when` from the end to the beginning. This is problematic since **short-circuit evaluation** is adopted by most engines, including PostgreSQL, Oracle, and SQL Server. Therefore, chances are users would express computation logic that would fail in later `when_thens` for tuples that **should have been** computed and bypassed.

Reply:
We can evaluate `case_when` from beginning to the end and use evaluate_selection for the when_expr so that we can omit the following computation of when expr of the record which when_expr is already true.